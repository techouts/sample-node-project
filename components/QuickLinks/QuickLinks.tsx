/* eslint-disable @next/next/no-img-element */
import CategoryInterface from "../../schemas/Categories";
import Title from "../../HOC/Title/Title";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import {
  CategoryImage,
  CategoryText,
  ListItemContainer,
} from "./QuickLinksStyle";
import { useEffect, useRef, useState } from "react";
import { ReplaceImage } from "../../utility/ReplaceImage";
import { widget_type, event_type } from "../../utility/GAConstants";
import { useMobileCheck } from "../../utility/isMobile";
import ViewEvent from "../../utility/viewEvent";
import triggerGAEvent from "../../utility/GaEvents";
import { ViewArray } from "../../utility/ViewEvenItemArray";
import { useInView } from "react-intersection-observer";
import { Cookies } from "react-cookie";
import useStorage from "../../utility/useStoarge";
import triggerGaViewPromotion from "../../utility/GAViewPromotion";

function QuickLinks({
  bgPadding,
  items,
  mobileItems,
  bgColor,
  title,
  id,
  __component,
  position,
  promotion_id,
  promotion_name,
}: CategoryInterface) {
  const isMobile = useMobileCheck();
  const [listItems, setListItems] = useState(isMobile ? mobileItems : items);
  const viewEventWrapper = useRef();
  const { ref, inView } = useInView({ triggerOnce: true, threshold: 0.1 });
  const [hasFiredEvent, setHasFiredEvent] = useState(false);

  const cookie = new Cookies();
  const { getItem } = useStorage();
  useEffect(() => {
    setListItems(isMobile ? mobileItems : items);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isMobile]);


  useEffect(() => {
    if (inView && !hasFiredEvent) {
      const datasLayer = {
        previous_screen_name: "na",
        visitor_type: cookie.get("accessToken") ? "loggedin" : "Guest",
        user_id: getItem("customer_ref", "local") || "na",
        creative_name: `${listItems
          ?.map((item: any) => `${item?.creative_name}-outer QL`)
          .join("|")}`,
        creative_slot: `${listItems
          ?.map((item: any) => `${item?.creative_slot}-outer QL`)
          .join("|")}`,
        items: listItems?.map((item) => ({
          promotion_id: `${promotion_id}-inner QL`,
          promotion_name: `${promotion_name}-inner QL`,
          creative_name: `${item?.creative_name}-inner QL`,
          creative_slot: `${item?.creative_slot}-inner QL`,
        })),
        // }
      };
      triggerGaViewPromotion("view_promotion", datasLayer);
      setHasFiredEvent(true);
    }
  }, [inView, hasFiredEvent, cookie, getItem, id, __component, listItems]);

  const handleCTAClick = (category: any, index: number) => {
    if (category?.isNewTab) {
      isMobile
        ? window.location.assign(category?.path)
        : window.open(category?.path);
    } else {
      window.location.assign(category?.path);
    }
    callClickEvent(category, index);
  };
  const callClickEvent = (category: any, index: number) => {
    const datasLayer = {
      previous_screen_name: "na",
      visitor_type: cookie.get("accessToken") ? "loggedin" : "Guest",
      user_id: getItem("customer_ref", "local") || "na",
      creative_name: `${category?.creative_name}-outer QL`,
      creative_slot: `${category?.creative_slot}-outer QL`,
      items: [
        {
          promotion_id: `${promotion_id}-inner QL`,
          promotion_name: `${promotion_name}-inner QL`,
          creative_name: `${category?.creative_name}-inner QL`,
          creative_slot: `${category?.creative_slot}-inner QL`,
        },
      ],
    };
    triggerGaViewPromotion("select_promotion", datasLayer);
  };
  return (
    <Box
      bgcolor={bgColor}
      p={isMobile ? "0 16px" : bgPadding}
      sx={{
        paddingTop: isMobile ? "12px" : "20px",
        paddingBottom: isMobile ? "12px" : "20px",
      }}
      className={__component + id}
      ref={viewEventWrapper}
    >
      {title && (
        <Box
          pb={isMobile ? "10px" : "18px"}
          sx={{
            paddingTop: isMobile ? "2px" : "4px",
            paddingBottom: isMobile ? "14px" : "24px",
          }}
        >
          <Title interface={{ title: title, bgColor: bgColor }} />
        </Box>
      )}
      <Grid
        container
        spacing={isMobile ? undefined : 0}
        rowGap={isMobile ? "10px" : 0}
        ref={ref}
      >
        {inView &&
          listItems?.map((category, idx) => (
            <Grid
              item
              xs={4}
              sm={12 / listItems?.length}
              md={12 / listItems?.length}
              lg={12 / listItems?.length}
              key={idx}
            >
              <ListItemContainer onClick={() => handleCTAClick(category, idx)}>
                <CategoryImage
                  src={`${ReplaceImage(category?.imageUrl)}`}
                  alt="Top Categories"
                  width="100%"
                />
                <CategoryText
                  container
                  direction="row"
                  justifyContent="center"
                  alignItems="center"
                >
                  {category?.text}
                </CategoryText>
              </ListItemContainer>
            </Grid>
          ))}
      </Grid>
    </Box>
  );
}

export default QuickLinks;