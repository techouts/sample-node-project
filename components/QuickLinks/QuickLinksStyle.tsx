import styled from "@emotion/styled";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";


export const CategoryImage = styled.img(() => ({
  cursor: "pointer",
  padding: "3px",
  "@media(min-width: 600px)": {
    padding: "6px",
  },
}));

export const CategoryText = styled(Grid)(() => ({
  fontWeight: "400",
  color: " #231f20",
  fontSize: "13px",
  cursor: "pointer",
  lineHeight: "135%",
  textAlign: "center",
  "@media(min-width: 600px)": {
    fontSize: "20px",
  },
}));

export const ListItemContainer = styled(Box)(() => ({
  padding: "0 8px",
  "@media(min-width: 768px)": {
    padding: "0 26px",
  },
}));
