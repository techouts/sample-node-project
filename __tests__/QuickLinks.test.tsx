import { render } from "@testing-library/react";
import QuickLinks from "../components/QuickLinks/QuickLinks";

const quickLinksData = {
  id: 1,
  __component: "widget.quick-links",
  bgColor: "#FFFFFF",
  bgPadding: "0% 5%",
  title: "TOP CATEGORIES FOR YOU",
  display: null,
  items: [
    {
      id: 1,
      imageUrl:
        "https://s3.ap-south-1.amazonaws.com/cmsimages.ssbeauty.in/eyeliners_desktop_b975059c2b.jpg?updated_at=2022-10-14T07:00:54.264Z",
      text: "Eyeliners",
      path: "/makeup/eyes/eyeliner/c/5?category_id=25",
      Item_name: "",
      Item_type: "",
      isNewTab: null,
    },
  ],
  mobileItems: [
    {
      id: 7,
      imageUrl:
        "https://s3.ap-south-1.amazonaws.com/cmsimages.ssbeauty.in/eyeliners_app_93915f1b3f.jpg?updated_at=2022-10-14T07:02:41.854Z",
      text: "Eyeliners",
      path: "/makeup/eyes/eyeliner/c/5?category_id=25",
      Item_name: "",
      Item_type: "",
      isNewTab: null,
    },
  ],
  position: 11,
};

describe("Quick Links", () => {
  it("renders Quick Links static page and creates a snapshot of the page", () => {
    const { container } = render(<QuickLinks {...quickLinksData} />);

    expect(container).toMatchSnapshot();
  });
});
