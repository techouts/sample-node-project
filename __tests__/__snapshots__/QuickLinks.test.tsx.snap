// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quick Guide renders Quick Links static page and creates a snapshot of the page 1`] = `
<div>
  <div
    class="widget.quick-links1 MuiBox-root css-kcnfl6"
  >
    <div
      class="MuiBox-root css-nqq86n"
    >
      <div
        class="MuiBox-root css-5p2tng"
      >
        TOP CATEGORIES FOR YOU
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-1ngrv4g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 css-1fazpqo-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-tov9kk"
        >
          <img
            alt="Top Categories"
            class="css-so1bna"
            src="https://cmsimages.ssbeauty.in/eyeliners_app_93915f1b3f.jpg"
            width="100%"
          />
          <div
            class="MuiGrid-root MuiGrid-container css-4mi3gg-MuiGrid-root"
          >
            Eyeliners
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Quick Links renders Quick Links static page and creates a snapshot of the page 1`] = `
<div>
  <div
    class="widget.quick-links1 MuiBox-root css-kcnfl6"
  >
    <div
      class="MuiBox-root css-nqq86n"
    >
      <div
        class="MuiBox-root css-5p2tng"
      >
        TOP CATEGORIES FOR YOU
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container css-1ngrv4g-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 MuiGrid-grid-lg-12 css-1fazpqo-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-tov9kk"
        >
          <img
            alt="Top Categories"
            class="css-so1bna"
            src="https://cmsimages.ssbeauty.in/eyeliners_app_93915f1b3f.jpg"
            width="100%"
          />
          <div
            class="MuiGrid-root MuiGrid-container css-4mi3gg-MuiGrid-root"
          >
            Eyeliners
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
